### ISSUE TITLE
how to implement cdist 

### ISSUE THREAD

**Issue by KukumavMozolo** [rusty1s/pytorch_sparse#378]
Hi there, thanks a million for this library.
I am trying to figure out how compute distances between every row(or column) of two matrices e.g. like `torch.cdist`.

Here is one way:
```python
def sparse_cdist(a: SparseTensor, b: SparseTensor):
        a_repeated = cat([a] * a.size(0), dim=0)
        b_repeated = cat(
            [cat([b[i, :]] * b.size(0), dim=0) for i in range(b.size(0))], dim=0
        )
        distances = sparse_distance(a_repeated, b_repeated)
        distances.requires_grad = False
    return distances.view((a.size(0), b.size(0)))
```
and another: 
```python

def sparse_cdist2(a: SparseTensor, b: SparseTensor):
    with torch.no_grad():
        distances = torch.zeros(
            (a.size(0), b.size(0)),
            device=a.device(),
            dtype=a.dtype(),
            requires_grad=False,
        )
        counter = 0
        for i in range(a.size(0)):
            for j in range(b.size(0)):
                distances[i, j] = sparse_distance(a[i, :], b[j, :])
    return distances
```
and another: 
```python

def sparse_cdist3(a: SparseTensor, b: SparseTensor):
    with torch.no_grad():
        distances = torch.zeros(
            (a.size(0), b.size(0)),
            device=a.device(),
            dtype=a.dtype(),
            requires_grad=False,
        )
        if a.size(0) <= b.size(0):
            for i in range(a.size(0)):
                idx = torch.tensor([i])
                idx = idx.expand(b.size(0))
                a_repeated = a.index_select(0,idx)
                distances[i, :] = sparse_distance(a_repeated, b)
        else:
            for j in range(b.size(0)):
                idx = torch.tensor([j])
                idx = idx.expand(a.size(0))
                b_repeated = b.index_select(0,idx)
                distances[:, j] = sparse_distance(a, b_repeated)
    return distances
```

where `sparse_distance` is defined as follows
```python
def sparse_distance(a: SparseTensor, b: SparseTensor):
    c = a + b.mul_nnz(torch.tensor(-1).to(device), "coo")
    c = c * c
    c = reduction(c, dim=1, reduce="sum")
    return torch.sqrt(c + 0.000000001)
```

The first one has the disadvantage that it creates huge matrices and eats a lot of memory, while the second one doesn't benefit from gpu parallelism. The third seems to be okish but maybe there is an even better solution? Ideally one would only load matrices a and b onto the gpu and only reserve additional memory for the result.
Maybe somebody has an idea how to do that with what is currently possible with torch sparse?
Or would it be necessary to write a specific cuda kernel for that?
Any suggestions are very welcome.

**Comment by rusty1s** [comment]
Interesting problem. The best way to implement this is indeed with a custom CUDA kernel, in which way you can drop the for-loop in option2/3 and avoid the memory blowup of option1. I think option3 is okay if you don't want to go into this rabbit hole of custom CUDA.

**Comment by KukumavMozolo** [comment]
I am currently trying to implement this in [cuda](https://github.com/KukumavMozolo/pytorch_sparse_addons),
at the moment working on the backward path and i am trying to follow the structure of how things are done in this project. Could it be that that in the bw path gradients are encoded as dense matrices? Because i only see one array there that seems to correspond to the gradients of the previous layer?

**Comment by KukumavMozolo** [comment]
Ok, i just found this: `grad_data[b * M * K + row * K + k]`, where b seems to be the batch dimension, k the number or rows and M the number of columns. So gradients are dense, is that correct?

**Comment by KukumavMozolo** [comment]
Another question: While one could compute the hole gradient matrix i think in practice this is not necessary in most cases. I for one would be just interested in the gradients for the vectors that are closest or e.g. farthest apart.
How are these cases usually handled efficiently?

**Comment by KukumavMozolo** [comment]
Regarding the example from pytorch_sparse abouve, i thought that adding this code 
```c++
SPARSE_API torch::Tensor spmm_sum(torch::optional<torch::Tensor> opt_row,
                       torch::Tensor rowptr, torch::Tensor col,
                       torch::optional<torch::Tensor> opt_value,
                       torch::optional<torch::Tensor> opt_colptr,
                       torch::optional<torch::Tensor> opt_csr2csc,
                       torch::Tensor mat) {
  auto value = opt_value.has_value() ? opt_value.value() : col;
  return SPMMSum::apply(opt_row, rowptr, col, value, opt_colptr, opt_csr2csc,
                        mat, opt_value.has_value())[0];
}
```
would apply an gradient function to the resultting tensor of `spmm_sum`. However that doesnt seem to be the case, when testing i am getting the following error:

**Comment by KukumavMozolo** [comment]
So i am currently busy implementing a [cuda](https://github.com/KukumavMozolo/pytorch_sparse_addons) kernel that computes the euclidean distance between two sets of matrix, e.g. similar to torch.cdist.

Unfortunately i ran into a problem when in the backward path when computing the local gradients:

Some definitions:

Let $`X \in R^{NxC}`$ and  $`Y \in R^{MxC}`$ then $`cdist(X,Y) \in R^{MxN}`$ and $`cdist(X,Y)_{i,j} = \sqrt{\sum_j (X_{i,j} - Y_{i,j})^2}`$

The derivative:
Here is the derivative of cdist with respect to some entry of X: $`\frac{\nabla cdist(X,Y)_{i,j}}{\nabla X_{i,n}} = \frac{X_{i,n} - Y_{i,n}}{\sqrt{\sum_j (X_{i,j} - Y_{i,j})^2}}`$

Then there can be 4 cases that need to be treated differently with sparse matrices(Ignoring that cdist is not differentiable at zero for now):
Case 1: $`X_{i,n} \neq 0  \, and \, Y_{i,n} \neq 0 \,\,\rightarrow \frac{\nabla cdist(X,Y)_{i,j}}{\nabla X_{i,n}} = \frac{X_{i,n} - Y_{i,n}}{\sqrt{\sum_j (X_{i,j} - Y_{i,j})^2}}`$
Case 2: $`X_{i,n} \neq 0 \, and \, Y_{i,n} = 0 \,\,\rightarrow \frac{\nabla cdist(X,Y)_{i,j}}{\nabla X_{i,n}} = \frac{X_{i,n} }{\sqrt{\sum_j (X_{i,j} - Y_{i,j})^2}}`$
Case 3: $`X_{i,n} = 0  \, and \, Y_{i,n} = 0 \,\,\rightarrow \frac{\nabla cdist(X,Y)_{i,j}}{\nabla X_{i,n}} = 0`$
Case 4: $`X_{i,n} = 0 \, and \, Y_{i,n} \neq 0 \,\,\rightarrow \frac{\nabla cdist(X,Y)_{i,j}}{\nabla X_{i,n}} = \frac{-Y_{i,n} }{\sqrt{\sum_j (X_{i,j} - Y_{i,j})^2}}`$

The problematic case is case 4!
Here, when computing the derivatives, the sparsity of X will change because where X was zero before the gradient would introduce a new non zero value. This requires to also change the size and values of rowPtr and colIndex together with the value array of the CSR Matrix. 
Hence my question: How could this be solved?

Here is the autograd boilerplate i am currently using:

```c++
using torch::autograd::AutogradContext;
using torch::autograd::Variable;
using torch::autograd::variable_list;

class SparseCdist : public torch::autograd::Function<SparseCdist> {
public: static variable_list forward(
    AutogradContext *ctx,
    torch::Tensor a_rowptr_data,
    torch::Tensor a_col_data,
    torch::Tensor a_value_data,
    torch::Tensor b_rowptr_data,
    torch::Tensor b_col_data,
    torch::Tensor b_value_data,
    int dim_a,
    int dim_b
    ) {
    auto out = sparse_cdist(a_rowptr_data, a_col_data, a_value_data, b_rowptr_data, b_col_data, b_value_data, dim_a, dim_b);
    ctx->saved_data["dim_a"] = dim_a;
    ctx->saved_data["dim_b"] = dim_b;
    ctx->save_for_backward({a_rowptr_data, a_col_data, a_value_data, b_rowptr_data, b_col_data, b_value_data, out});
    return {out};
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_outs) {
    auto dim_a = ctx->saved_data["dim_a"].toInt();
    auto dim_b = ctx->saved_data["dim_b"].toInt();
    auto grad_out = grad_outs[0];
    auto saved = ctx->get_saved_variables();
    auto a_rowptr_data = saved[0], a_col_data = saved[1], a_value_data = saved[2], b_rowptr_data = saved[3],
         b_col_data = saved[4], b_value_data = saved[5], distance = saved[5];

    auto grad_value_a = Variable();
    if (torch::autograd::any_variable_requires_grad({a_value_data})){
      std::cout << "grad_outs is: " << grad_out;
      grad_value_a = sparse_bw_cdist(a_rowptr_data, a_col_data, a_value_data, b_rowptr_data, b_col_data, b_value_data, grad_out, distance, dim_a, dim_b);
    }
    
    auto grad_value_b = Variable();
    if (torch::autograd::any_variable_requires_grad({b_value_data})){
      std::cout << "grad_outs is: " << grad_out;
      grad_value_b = sparse_bw_cdist(b_rowptr_data, b_col_data, b_value_data,a_rowptr_data, a_col_data, a_value_data, grad_out, distance, dim_b, dim_a);
    }    
    return {Variable(), Variable(), grad_value_a,
            Variable(), Variable(), grad_value_b, Variable(), Variable()};
  }
};
```

What i would want to do is overwrite `a_rowptr_data` and `a_col_data` but it is unclear to me how this can be done using this AutogradContext concept.
Help would be much appreciated, also if this starts to work would there be interest to integrate `cdist` it into this project?

**Comment by KukumavMozolo** [comment]
Maybe one workaround is to explicitly introduce zeros to X where X is implicitly is zero and Y is none zero?
